README
FlatFile-DataStore-Lite
=======================

INSTALLATION

To install this module type the following:

    perl Makefile.PL
    make
    make test
    make install

DEPENDENCIES

This module requires these other modules and libraries:

   URI
   URI::Escape
   Data::Omap
   Math::Int2Base

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2009 by Brad Baxter

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension FlatFile::DataStore::Lite.

0.01  Fri Mar 20 15:27:52 2009
    - original version; created by h2xs 1.23 with options
        -AX --skip-exporter --use-new-tests -n FlatFile::DataStore::Lite -b 5.8.3



lib/FlatFile/DataStore/Lite/Toc.pm
#---------------------------------------------------------------------
  package FlatFile::DataStore::Lite::Toc;
#---------------------------------------------------------------------

=head1 NAME

FlatFile::DataStore::Lite::Toc - Perl module that implements a
lightweight flat file data store TOC (table of contents) class.

=head1 SYNOPSYS

 use FlatFile::DataStore::Lite::Toc;
 my $toc;

 $toc = FlatFile::DataStore::Lite::Toc->new( { int => 10,
     datastore => $datastore_obj } );

 # or

 $toc = FlatFile::DataStore::Lite::Toc->new( { num => "A",
     datastore => $datastore_obj } );

=head1 DESCRIPTION

FlatFile::DataStore::Lite::Toc is a Perl module that implements a
lightweight flat file data store TOC (table of contents) class.

This module is used by FlatFile::DataStore::Lite.  You will likely
never call any of it's methods yourself.

=head1 VERSION

FlatFile::DataStore::Lite::Toc version 0.01

=cut

our $VERSION = '0.01';

use 5.008003;
use strict;
use warnings;

use File::Path;
use Carp;
use Math::Int2Base qw( base_chars int2base base2int );

my %Attrs = qw(
    datastore 1
    datafnum  1
    keyfnum   1
    tocfnum   1
    numrecs   1
    keynum    1
    );

#---------------------------------------------------------------------

=head1 CLASS METHODS

=head2 FlatFile::DataStore::Lite::Toc->new( $parms )

Constructs a new FlatFile::DataStore::Lite::Toc object.

The parm C<$parms> is a hash reference containing key/value pairs to
populate the record string.  Three keys are recognized:

 - datastore, data store object (required) and one of:
 - int, data file number as integer, will load object from tocfile
   or
 - num, data file number as number in number base, will load from tocfile

An C<int> or C<num> of 0 will load the first (totals) line from tocfile

=cut

sub new {
    my( $class, $parms ) = @_;

    my $self = bless {}, $class;

    $self->init( $parms ) if $parms;
    return $self;
}


#---------------------------------------------------------------------
# init(), called by new() to parse the parms

sub init {
    my( $self, $parms ) = @_;

    my $ds = $parms->{'datastore'} || croak "Missing datastore";
    $self->datastore( $ds );

    my $datafint;
    if(    defined( my $int = $parms->{'int'} ) ) {
        $datafint = $int; }
    elsif( defined( my $num = $parms->{'num'} ) ) {
        $datafint = base2int $num, $ds->fnumbase; }
    else {
        croak qq/Missing parms 'int' or 'num'./; }

    my $string = $self->read_toc( $datafint );

    unless( $string ) {
        $self->datafnum( $datafint );
        $self->tocfnum(  $self->toc_getfnum( $datafint ) );
        $self->keynum(   $datafint == 0? -1: 0 );
        $self->keyfnum(  0 );
        $self->numrecs(  0 );
        return $self;
    }

    my $fnumbase  = $ds->fnumbase;
    my $keybase   = $ds->keybase;

    my $recsep = $ds->recsep;
    $string =~ s/$recsep$//;  # chompish
    $self->string( $string );

    my @fields = split " ", $string;
    my $i = 0;
    $self->$_( base2int $fields[ $i++ ], $fnumbase )
        for qw( datafnum keyfnum tocfnum );
    $self->$_( base2int $fields[ $i++ ], $keybase )
        for qw( numrecs keynum );

    return $self;
}

#---------------------------------------------------------------------

=head1 OBJECT METHODS

=cut

#---------------------------------------------------------------------
sub to_string {
    my( $self ) = @_;

    my $ds = $self->datastore;

    my $fnumbase  = $ds->fnumbase;
    my $fnumlen   = $ds->fnumlen;
    my $keybase   = $ds->keybase;
    my $keylen    = $ds->keylen;

    my @fields;
    push @fields, int2base $self->$_(), $fnumbase, $fnumlen
        for qw( datafnum keyfnum tocfnum );
    push @fields, int2base $self->$_(), $keybase, $keylen
        for qw( numrecs keynum );

    return join( " " => @fields ) . $ds->recsep;
}

#---------------------------------------------------------------------
sub read_toc {
    my( $self, $fint ) = @_;

    my $ds = $self->datastore;

    my $tocfile = $self->tocfile( $fint );
    return unless -e $tocfile;

    my $tocfh  = $ds->locked_for_read( $tocfile );
    my $toclen = $ds->toclen;

    my $seekpos;
    if( my $tocmax = $ds->tocmax ) {
        my $skip = int( $fint / $tocmax );
        $seekpos = $toclen * ( $fint - ( $skip * $tocmax ) ); }
    else {
        $seekpos = $toclen * $fint; }

    return $ds->read_bytes( $tocfh, $seekpos, $toclen );
}

#---------------------------------------------------------------------
sub write_toc {
    my( $self, $fint ) = @_;

    my $ds = $self->datastore;

    my $tocfh   = $ds->locked_for_write( $self->tocfile( $fint ) );
    my $toclen  = $ds->toclen;

    my $seekpos;
    if( my $tocmax = $ds->tocmax ) {
        my $skip = int( $fint / $tocmax );
        $seekpos = $toclen * ( $fint - ( $skip * $tocmax ) ); }
    else {
        $seekpos = $toclen * $fint; }

    return $ds->write_bytes( $tocfh, $seekpos, $self->to_string );
}

#---------------------------------------------------------------------
# toc_getfnum(), called by tocfile() and init()
sub toc_getfnum {
    my( $self, $fint ) = @_;

    my $ds = $self->datastore;

    # get toc file number based on tocmax and fint
    my $tocfint;

    my $tocmax = $ds->tocmax;
    if( $tocmax ) { $tocfint = int( $fint / $tocmax ) + 1 }
    else          { $tocfint = 1                          }

    my $fnumlen  = $ds->fnumlen;
    my $fnumbase = $ds->fnumbase;
    my $tocfnum = int2base $tocfint, $fnumbase, $fnumlen;
    croak qq/Database exceeds configured size (tocfnum: "$tocfnum" too long)/
        if length $tocfnum > $fnumlen;
    return( $tocfint, $tocfnum ) if wantarray;
    return $tocfint;
}

#---------------------------------------------------------------------
sub tocfile {
    my( $self, $fint ) = @_;

    my $ds = $self->datastore;

    my $name = $ds->name;

    my( $tocfint, $tocfnum ) = $self->toc_getfnum( $fint );
    my $tocfile = $name . ( $ds->tocmax? ".$tocfnum": "") . ".toc";

    # get toc path based on dirlev, dirmax, and toc file number
    if( my $dirlev = $ds->dirlev ) {
        my $fnumlen  = $ds->fnumlen;
        my $fnumbase = $ds->fnumbase;
        my $dirmax   = $ds->dirmax;
        my $path     = "";
        my $this     = $tocfint;
        for( 1 .. $dirlev ) {
            my $dirint = $dirmax? (int( ( $this - 1 ) / $dirmax ) + 1): 1;
            my $dirnum = int2base $dirint, $fnumbase, $fnumlen;
            $path = $path? "$dirnum/$path": $dirnum;
            $this = $dirint;
        }
        $path = "$name/toc$path";
        mkpath( $path ) unless -d $path;
        $tocfile = "$path/$tocfile";
    }

    return $ds->dir . "/$tocfile";
}

#---------------------------------------------------------------------

=head1 OBJECT METHODS: Accessors

The following read/write methods set and return their respective
attribute values if C<$value> is given.  Otherwise, they just return
the value.

 $record->datastore( [$value] )
 $record->string(    [$value] )

The following methods expect an integer parm and return an integer
value (even though these are stored in the tocfile as numbers in their
respective bases).

 $record->datafnum( [$value] )
 $record->keyfnum(  [$value] )
 $record->tocfnum(  [$value] )
 $record->numrecs(  [$value] )
 $record->keynum(   [$value] )

=cut

sub datastore {for($_[0]->{datastore} ){$_=$_[1]if@_>1;return$_}}
sub string    {for($_[0]->{string}    ){$_=$_[1]if@_>1;return$_}}

sub datafnum  {for($_[0]->{datafnum}  ){$_=$_[1]if@_>1;return$_}}
sub keyfnum   {for($_[0]->{keyfnum}   ){$_=$_[1]if@_>1;return$_}}
sub tocfnum   {for($_[0]->{tocfnum}   ){$_=$_[1]if@_>1;return$_}}
sub numrecs   {for($_[0]->{numrecs}   ){$_=$_[1]if@_>1;return$_}}
sub keynum    {for($_[0]->{keynum}    ){$_=$_[1]if@_>1;return$_}}

__END__

=head1 AUTHOR

Brad Baxter, E<lt>bbaxter@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Brad Baxter

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut



lib/FlatFile/DataStore/Lite/Record.pm
#---------------------------------------------------------------------
  package FlatFile::DataStore::Lite::Record;
#---------------------------------------------------------------------

=head1 NAME

FlatFile::DataStore::Lite::Record - Perl module that implements a
lightweight flat file data store record class.

=head1 SYNOPSYS

 # first, create a preamble object

 use FlatFile::DataStore::Lite::Preamble;

 my $preamble = FlatFile::DataStore::Lite::Preamble->new( {
     indicator => $indicator,  # single-character crud flag
     keynum    => $keynum,     # record sequence number (integer)
     reclen    => $reclen,     # record length (integer)
     thisfnum  => $fnum,       # file number (in base format)
     thisseek  => $datapos,    # seek position (integer)
     user      => $user_data,  # pre-formatted user-defined data
     } );

 # then create a record object with the preamble contained in it

 use FlatFile::DataStore::Lite::Record;

 my $record = FlatFile::DataStore::Lite::Record->new( {
     preamble => $preamble,                 # i.e., a preamble object
     data     => "This is a test record.",  # actual record data
     } );

=head1 DESCRIPTION

FlatFile::DataStore::Lite::Record is a Perl module that implements a
lightweight flat file data store record class.  This class defines
objects used by FlatFile::DataStore::Lite.  You will likely not ever
call new() yourself, (FlatFile::DataStore::Lite::create() would, e.g.,
do that) but you will likely call the accessors.

=head1 VERSION

FlatFile::DataStore::Lite::Record version 0.01

=cut

our $VERSION = '0.01';

use 5.008003;
use strict;
use warnings;

use Carp;

my %Attrs = qw(
    preamble  1
    data      1
    );

#---------------------------------------------------------------------

=head1 CLASS METHODS

=head2 FlatFile::DataStore::Lite::Record->new( $parms )

Constructs a new FlatFile::DataStore::Lite::Record object.

The parm C<$parms> is a hash reference containing key/value pairs to
populate the record string.  Two keys are recognized:

 - preamble, i.e., a FlatFile::DataStore::Lite::Preamble object
 - data,     the actual record data

The record data is stored as a scalar reference.

=cut

sub new {
    my( $class, $parms ) = @_;

    my $self = bless {}, $class;

    $self->init( $parms ) if $parms;
    return $self;
}


#---------------------------------------------------------------------
# init(), called by new() to parse the parms

sub init {
    my( $self, $parms ) = @_;

    # want to store record data as a scalar reference
    for( $parms->{'data'} ) {
        if( defined ) {
            if( ref eq 'SCALAR' ) { $self->data( $_  ) }
            else                  { $self->data( \$_ ) }
        }
        else                      { $self->data( \"" ) }
    }

    if( my $preamble = $parms->{'preamble'} ) {
        $self->preamble( $preamble );
    }
    
    return $self;
}

#---------------------------------------------------------------------

=head1 OBJECT METHODS: ACCESSORS

The following read/write methods set and return their respective
attribute values if C<$value> is given.  Otherwise, they just return
the value.

 $record->data(     [$value] ); # actual record data
 $record->preamble( [$value] ); # FlatFile::DataStore::Lite::Preamble object

=cut

sub data     {for($_[0]->{data}    ){$_=$_[1]if@_>1;return$_}}
sub preamble {for($_[0]->{preamble}){$_=$_[1]if@_>1;return$_}}

=pod

The following read-only methods just return their respective values.
The values all come from the record's contained preamble object.

 $record->user()
 $record->string()
 $record->indicator()
 $record->keynum()
 $record->reclen()
 $record->thisfnum()
 $record->thisseek()

=cut

sub user {for($_[0]->preamble()){defined&&return$_->user()}}

sub string    {$_[0]->preamble()->string()   }
sub indicator {$_[0]->preamble()->indicator()}
sub keynum    {$_[0]->preamble()->keynum()   }
sub reclen    {$_[0]->preamble()->reclen()   }
sub thisfnum  {$_[0]->preamble()->thisfnum() }
sub thisseek  {$_[0]->preamble()->thisseek() }

__END__

=head1 AUTHOR

Brad Baxter, E<lt>bbaxter@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Brad Baxter

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut



lib/FlatFile/DataStore/Lite/Preamble.pm
#---------------------------------------------------------------------
  package FlatFile::DataStore::Lite::Preamble;
#---------------------------------------------------------------------

=head1 NAME

FlatFile::DataStore::Lite::Preamble - Perl module that implements a
lightweight flat file data store preamble class.

=head1 SYNOPSYS

 use FlatFile::DataStore::Lite::Preamble;

 my $preamble = FlatFile::DataStore::Lite::Preamble->new( {
     indicator => $indicator,  # single-character crud flag
     keynum    => $keynum,     # record sequence number (integer)
     reclen    => $reclen,     # record length (integer)
     thisfnum  => $fnum,       # file number (in base format)
     thisseek  => $datapos,    # seek position (integer)
     user      => $user_data,  # pre-formatted user-defined data
     } );

 my $string = $preamble->string(); # e.g.,

 # new preamble from existing preamble string
 # e.g., something like "=000I00Tw1001XNTest "

 my $clone = FlatFile::DataStore::Lite::Preamble->new( { string => $string } );

=head1 DESCRIPTION

FlatFile::DataStore::Lite::Preamble - Perl module that implements a
lightweight flat file data store preamble class.  This class defines
objects used by FlatFile::DataStore::Lite::Record and
FlatFile::DataStore::Lite.  So you will probably not ever call new()
yourself, but you might call some of the accessors either directly or
via a FF::DS::Record object;

A "preamble" is a string of fixed-length fields that precedes every
record in a FlatFile::DataStore::Lite data file.  In addition, this
string constitutes the entry in the data store key file for each
current record.

=head1 VERSION

FlatFile::DataStore::Lite::Preamble version 0.01

=cut

our $VERSION = '0.01';

use 5.008003;
use strict;
use warnings;

use Carp;
use Math::Int2Base qw( base_chars int2base base2int );
use Data::Omap qw( :ALL );

my %Generated = qw(
    string      1
    );

my %Attrs = ( %Generated, qw(
    indicator 1
    keynum    1
    reclen    1
    thisfnum  1
    thisseek  1
    user      1
    ) );

my $Ascii_chars = qr/^[ -~]+$/;

#---------------------------------------------------------------------

=head1 CLASS METHODS

=head2 FlatFile::DataStore::Lite::Preamble->new( $parms )

Constructs a new FlatFile::DataStore::Lite::Preamble object.

The parm C<$parms> is a hash reference containing key/value pairs to
populate the preamble string.  If there is a C<$parms->{'string'}>
value, it will be parsed into fields and the resulting key/value pairs
will replace the C<$parms> hash reference.

=cut

sub new {
    my( $class, $parms ) = @_;

    my $self = bless {}, $class;

    $self->init( $parms ) if $parms;
    return $self;
}

#---------------------------------------------------------------------
# init(), called by new() to parse the parms

sub init {
    my( $self, $parms ) = @_;

    my $datastore = $parms->{'datastore'} || croak "Missing datastore";
    if( my $string = $parms->{'string'} ) {
        $parms = $datastore->burst_preamble( $string );
    }

    my $crud   = $datastore->crud();
    my $create = $crud->{'create'};
    my $update = $crud->{'update'};
    my $delete = $crud->{'delete'};

    my $indicator = $parms->{'indicator'} || croak "Missing indicator";

    $self->indicator( $indicator );

    my $string = "";
    for my $href ( $datastore->specs() ) {  # each field is href of aref
        my( $field, $aref )     = %$href;
        my( $pos, $len, $parm ) = @$aref;
        my $value               = $parms->{ $field };

        for( $field ) {
            if( /indicator/ ) {
                croak qq'Missing value for "$_"' unless defined $value;
                croak qq'Invalid value for "$_" ($value)' unless length $value == $len;

                $self->{ $_ } = $value;
                $string      .= $value;
            }
            elsif( /user/ ) {
                croak qq'Missing value for "$_"' unless defined $value;
                croak qq'Invalid value for "$_" ($value)' unless $value =~ $Ascii_chars;

                my $try = sprintf "%-${len}s", $value;  # pads with blanks
                croak qq'Value of "$_" ($try) too long' if length $try > $len;

                $self->{ $_ } = $value;
                $string      .= $try;
            }
            elsif( not defined $value ) {
                if( (/keynum|reclen|thisfnum|thisseek/ ) ) {
                    croak qq'Missing value for "$_"';
                }
                $string .= "-" x $len;  # string of '-' for null
            }
            else {
                my $try = sprintf "%0${len}s", /fnum/? $value: int2base( $value, $parm );
                croak qq'Value of "$_" ($try) too long' if length $try > $len;

                $self->{ $_ } = /fnum/? $try: 0+$value;
                $string      .= $try;
            }
        }
    }

    croak qq'Something is wrong with preamble string: "$string"'
        unless $string =~ $datastore->regx();
    
    $self->string( $string );

    return $self;
}

#---------------------------------------------------------------------

=head1 OBJECT METHODS: ACCESSORS

The following methods set and return their respective attribute values
if C<$value> is given.  Otherwise, they just return the value.

 $preamble->string(    [$value] ); # full preamble string
 $preamble->indicator( [$value] ); # single-character crud indicator
 $preamble->keynum(    [$value] ); # record sequence number (integer)
 $preamble->reclen(    [$value] ); # record length (integer)
 $preamble->thisfnum(  [$value] ); # file number (in base format)
 $preamble->thisseek(  [$value] ); # seek position (integer)
 $preamble->user(      [$value] ); # pre-formatted user-defined data

Note: the class code uses these accessors to set values in the object
as it is assembling the preamble string in new().  Unless you have a
really good reason, you should not set these values yourself (outside
of a call to new()).  For example: setting the user data with user()
will I<not> change the user data in the C<string> attribute.

In other words, even though these are read/write accessors, you should
only use them for reading.

=cut

sub string    {for($_[0]->{string}    ){$_=$_[1]if@_>1;return$_}}
sub indicator {for($_[0]->{indicator} ){$_=$_[1]if@_>1;return$_}}
sub user      {for($_[0]->{user}      ){$_=$_[1]if@_>1;return$_}}

sub keynum    {for($_[0]->{keynum}    ){$_=0+$_[1]if@_>1;return$_}}
sub reclen    {for($_[0]->{reclen}    ){$_=0+$_[1]if@_>1;return$_}}
sub thisfnum  {for($_[0]->{thisfnum}  ){$_=  $_[1]if@_>1;return$_}}
sub thisseek  {for($_[0]->{thisseek}  ){$_=0+$_[1]if@_>1;return$_}}

__END__

=head1 AUTHOR

Brad Baxter, E<lt>bbaxter@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Brad Baxter

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut

